api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,570 @@
+				ws.Cipher(buf, header.Mask, offset)
+			}
+			offset += chunk
+			c.receiveBuffer.grow(uint64(chunk))
+			remaining -= int64(chunk)
+		}
+
+		c.cond.Broadcast()
+
+	}
+	c.mu.Unlock()
+
+	return nil
+}
+
+// readFramePartialPayloadLocked reads a portion of the frame payload of len(buf) into the provided buffer.
+// It assumes that the connection mutex is already locked and unlocks it during the read
+// to avoid blocking other write operations. After reading, it re-locks the mutex.
+// It's the caller's responsibility to ensure that the buffer size does not exceed
+// the remaining payload length.
+// It's safe to unlock the mutex here because the readLoop is the only
+// goroutine that reads from the underlying connection and advances the buffer's .end field.
+// The conn.Read call only updates the .start field, which means it can only
+// increase the available space in the buffer, never reduce it—so there’s no risk of race conditions.
+func (c *Conn) readFramePartialPayloadLocked(buf []byte) error {
+	// Unlock the mutex while reading to avoid blocking other operations.
+	c.mu.Unlock()
+	defer c.mu.Lock()
+	_, err := io.ReadFull(c.underlyingConn, buf)
+	return err
+}
+
+// readControlPayload reads the payload of a control frame (Ping, Pong, Close).
+// It ensures that the payload length does not exceed the maximum allowed size
+// for control frames (125 bytes) and returns an error if it does.
+// The function reads the entire payload into a byte slice and returns it.
+func (c *Conn) readControlPayload(header ws.Header) ([]byte, error) {
+	if header.Length < 0 || header.Length > ws.MaxControlFramePayloadSize {
+		return nil, ws.ErrProtocolControlPayloadOverflow
+	}
+
+	payload := make([]byte, header.Length)
+	n, err := io.ReadFull(c.underlyingConn, payload)
+	return payload[:n], err
+}
+
+// drainFramePayload discards the payload of a WebSocket frame by reading
+// and ignoring it. This is used for frames where the payload is not needed,
+// such as control frames that are not processed further.
+func (c *Conn) drainFramePayload(header ws.Header) error {
+	// Discard the frame payload by reading it without processing.
+	_, err := io.CopyN(io.Discard, c.underlyingConn, header.Length)
+	return err
+}
+
+// writeLoop continuously sends WebSocket frames from the send buffer to the
+// underlying connection. It manages transmission of binary data, ping/pong
+// responses, and close frames as required. The loop ends when the connection
+// is closed locally or remotely, ensuring a close frame is sent if applicable.
+// It also releases the read loop by setting a read deadline when needed.
+func (c *Conn) writeLoop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// unblockReaderOnErr is a helper function that sets a past read deadline
+	// on the underlying connection to unblock the read loop in case of errors
+	// during writing. It also broadcasts the condition variable to wake up
+	// any waiting goroutines.
+	unblockReaderOnErr := func() {
```

Comment:
A reminder that this requires holding on the managedConn lock.

```suggestion
	unblockReaderOnErrLocked := func() {
```